### PR TITLE
feat: refactoring LruCacheLayer with list_with_metakey and concurrent_stat_in_list

### LINKED ISSUE #3059: Consider refactoring LruCacheLayer with `list_with_metakey` and `concurrent_stat_in_list`
Consider refactoring it with [list_with_metakey](https://github.com/apache/incubator-opendal/blob/035b3dc48a26ac1350c0ba12872f9d97d602840c/core/src/docs/rfcs/2779_list_with_metakey.md) and [concurrent_stat_in_list](https://github.com/apache/incubator-opendal/blob/035b3dc48a26ac1350c0ba12872f9d97d602840c/core/src/docs/rfcs/3574_concurrent_stat_in_list.md)

_Originally posted by @WenyXu in https://github.com/GreptimeTeam/greptimedb/pull/3058#discussion_r1438831103_

https://github.com/GreptimeTeam/greptimedb/blob/2f98fa0d97195c4818a9db52032924700ca49577/src/object-store/src/layers/lru_cache/read_cache.rs#L135-L157

### PR DESCRIPTION
**Pull request by ozewr**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)
close #3059 
## What's changed and what's your intention?
 - build `Operator` from `file_cache` in `recover_cache` function
 - use `list_with_metakey` and `concurrent_stat_in_list`  instead of `list`
## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

### CODE REVIEW

**Comment by ozewr** [comment]
Is it Ok? I used clone . But if I don't use clone ,I cannot use  `list_with_metakey` and `concurrent_stat_in_list`.

**Comment by WenyXu** [comment]
> Is it Ok? I used clone . But if I don't use clone ,I cannot use `list_with_metakey` and `concurrent_stat_in_list`. Perhaps not do ant change would be better?

Hi @ozewr, we can retrieve the content lenght in `list_with_metakey` too.

```rust
let entries: Vec<Entry> = op
  .list_with("dir")
  .metakey(Metakey::ContentLength | Metakey::ContentType).await?;
```

then we can avoid to call the `stat` function:
```rust
let size = {
    let stat = self.file_cache.stat(read_key, OpStat::default()).await?;
    stat.into_metadata().content_length()
};
```

File: src/object-store/src/layers/lru_cache/read_cache.rs [hunk 1]
```diff
@@ -142,18 +142,20 @@ impl<C: Access> ReadCache<C> {
...
     /// Return entry count and total approximate entry size in bytes.
     pub(crate) async fn recover_cache(&self) -> Result<(u64, u64)> {
-        let (_, mut pager) = self.file_cache.list("/", OpList::default()).await?;
-
-        while let Some(entry) = pager.next().await? {
+        let op = OperatorBuilder::new(self.file_cache.clone()).finish();
+        let mut entries = op
+            .list_with("/")
+            .metakey(Metakey::ContentLength | Metakey::ContentType)
+            .concurrent(1)
```
Comment (WenyXu) [inline src/object-store/src/layers/lru_cache/read_cache.rs:149]: nit: let's use the const, and how about setting the `concurrent` to 8 or larger?

**Review by WenyXu — APPROVED** [review]
Rest LGTM

File: src/object-store/src/layers/lru_cache/read_cache.rs [hunk 2]
```diff
@@ -18,15 +18,15 @@ use common_telemetry::debug;
...
+use opendal::raw::oio::{Read, Reader, Write};
+use opendal::raw::{Access, OpDelete, OpRead, OpStat, OpWrite, RpRead};
+use opendal::{Error as OpendalError, ErrorKind, Metakey, OperatorBuilder, Result};
 
 use crate::metrics::{
     OBJECT_STORE_LRU_CACHE_BYTES, OBJECT_STORE_LRU_CACHE_ENTRIES, OBJECT_STORE_LRU_CACHE_HIT,
     OBJECT_STORE_LRU_CACHE_MISS, OBJECT_STORE_READ_ERROR,
 };
-
+const RECOVER_CACHE_CONCURRENT: usize = 8;
```
Comment (evenyag) [inline src/object-store/src/layers/lru_cache/read_cache.rs:29]:
```suggestion

const RECOVER_CACHE_CONCURRENT: usize = 8;

```

**Review by evenyag — APPROVED** [review]
Others LGTM

File: src/object-store/src/layers/lru_cache/read_cache.rs [hunk 3]
```diff
@@ -18,15 +18,17 @@ use common_telemetry::debug;
...
+use opendal::raw::oio::{Read, Reader, Write};
+use opendal::raw::{Access, OpDelete, OpRead, OpStat, OpWrite, RpRead};
+use opendal::{Error as OpendalError, ErrorKind, Metakey, OperatorBuilder, Result};
 
 use crate::metrics::{
     OBJECT_STORE_LRU_CACHE_BYTES, OBJECT_STORE_LRU_CACHE_ENTRIES, OBJECT_STORE_LRU_CACHE_HIT,
     OBJECT_STORE_LRU_CACHE_MISS, OBJECT_STORE_READ_ERROR,
 };
 
+const RECOVER_CACHE_CONCURRENT: usize = 8;
```
Comment (WenyXu) [inline src/object-store/src/layers/lru_cache/read_cache.rs:30]:
```suggestion
const RECOVER_CACHE_LIST_CONCURRENT: usize = 8;
```

File: src/object-store/src/layers/lru_cache/read_cache.rs [hunk 4]
```diff
@@ -142,19 +144,16 @@ impl<C: Access> ReadCache<C> {
...
     /// Return entry count and total approximate entry size in bytes.
     pub(crate) async fn recover_cache(&self) -> Result<(u64, u64)> {
-        let (_, mut pager) = self.file_cache.list("/", OpList::default()).await?;
-
-        while let Some(entry) = pager.next().await? {
+        let op = OperatorBuilder::new(self.file_cache.clone()).finish();
+        let mut entries = op
+            .list_with("/")
+            .metakey(Metakey::ContentLength | Metakey::ContentType)
+            .concurrent(RECOVER_CACHE_CONCURRENT)
```
Comment (WenyXu) [inline src/object-store/src/layers/lru_cache/read_cache.rs:151]:
```suggestion
            .concurrent(RECOVER_CACHE_LIST_CONCURRENT)
```

**Review by WenyXu — COMMENTED** [review]
some nitpicks:
- Always make sure name is unambiguity

**Pull request merged**